src/librustc_mir/const_eval.rs

```diff
@@ -536,9 +538,17 @@ fn validate_and_turn_into_const<'a, 'tcx>(
             )?;
         }
         // Now that we validated, turn this into a proper constant.
-        let def_id = cid.instance.def.def_id();
-        let normalize = tcx.is_static(def_id).is_none() && cid.promoted.is_none();
-        op_to_const(&ecx, op, normalize)
+        let val = match op.layout.abi {
+            layout::Abi::Scalar(..) => ConstValue::Scalar(ecx.read_immediate(op)?.to_scalar()?),
```

Comment:
This is the key point where we decide what can get "cached" in `val`, right? Might be worth adding a comment explaining what's going on.